File: sjf_non_preemptive.php
<!DOCTYPE html>
<html>
<head>
  <title>Non-preemptive Shortest Job First</title>
  <style>
    .container {
      max-width: 1150px;
      margin: 0 auto;
      padding: 20px;
      text-align: center;
    }

    h1 {
      margin-bottom: 20px;
    }

    textarea {
      width: 100%;
      resize: vertical;
      margin-bottom: 10px;
    }

    input[type="submit"] {
      padding: 10px 20px;
      background-color: #0047AB;
      color: white;
      border: none;
      cursor: pointer;
    }

    table {
      margin-top: 20px;
      width: 100%;
      border-collapse: collapse;
    }

    th, td {
      padding: 8px;
      text-align: left;
      border-bottom: 1px solid #ddd;
    }

    th {
      background-color: #0047AB;
      color: white;
    }

    .gantt {
      margin-top: 20px;
      display: flex;
      align-items: center;
    }

    .gantt-bar {
      height: 20px;
      background-color: #0047AB;
      margin-right: 4px;
      position: relative;
    }

    .gantt-text {
      position: absolute;
      top: 50%;
      left: 50%;
      transform: translate(-50%, -50%);
      color: white;
    }

    .gantt-timeline {
      display: flex;
      justify-content: center;
      margin-top: 8px;
    }

    .gantt-timeline-item {
      margin-right: 20px;
    }

    .time {
      display: flex;
      justify-content: center;
      margin-top: 8px;
    }

    .time-item {
      margin-right: 20px;
    }
  </style>
</head>
<body>
  <div class="container">
    <h1>Non-preemptive Shortest Job First</h1>
    <form action="<?php echo htmlspecialchars($_SERVER["PHP_SELF"]); ?>" method="post">
      <label for="processes">Enter process information:</label>
      <textarea id="processes" name="processes" rows="4" cols="50" placeholder="Enter processes in the format: ProcessName ArrivalTime BurstTime"><?php echo isset($_POST["processes"]) ? htmlspecialchars($_POST["processes"]) : ""; ?></textarea>
      <br>
      <input type="submit" value="Run">
    </form>

    <?php
    if ($_SERVER["REQUEST_METHOD"] == "POST") {
      // Get the input processes
      $processesInput = $_POST["processes"];

      // Parse the input into an array of processes
      $processesArray = explode("\n", $processesInput);
      $processes = array();

      foreach ($processesArray as $processString) {
        $processData = explode(" ", $processString);
        $processes[] = array(
          "name" => $processData[0],
          "arrivalTime" => intval($processData[1]),
          "burstTime" => intval($processData[2])
        );
      }

      // Sort processes based on arrival time
      usort($processes, function ($a, $b) {
        return $a["arrivalTime"] - $b["arrivalTime"];
      });

      // Algorithm
      $completionTimes = array();
      $turnaroundTimes = array();
      $waitingTimes = array();
      $ganttChart = array();
      $ganttStartTimes = array();
      $finalCompletedProcess = 0;

      $time = 0;
      $completedProcesses = 0;

      while ($completedProcesses < count($processes)) {
        $nextProcess = null;
        $shortestBurst = PHP_INT_MAX;

        foreach ($processes as $process) {
          if ($process["arrivalTime"] <= $time && $process["burstTime"] < $shortestBurst && !isset($completionTimes[$process["name"]]) && $process["burstTime"] != 0) {
            $nextProcess = $process;
            $shortestBurst = $process["burstTime"];
          }
        }

        if ($nextProcess == null) {
          $time++;
          continue;
        }

        $processName = $nextProcess["name"];
        $completionTimes[$processName] = $time + $nextProcess["burstTime"];
        $turnaroundTimes[$processName] = $completionTimes[$processName] - $nextProcess["arrivalTime"];
        $waitingTimes[$processName] = $turnaroundTimes[$processName] - $nextProcess["burstTime"];
        $completedProcesses++;

        for ($i = $time; $i < $completionTimes[$processName]; $i++) {
          if (!isset($ganttChart[$i])) {
            $ganttChart[$i] = $processName;
          }
        }

        $time = $completionTimes[$processName];
      }

      // Output the completion times, turnaround times, and waiting times
      echo "<h2>Table</h2>";
      echo "<table>";
      echo "<tr><th>Process</th><th>Arrival Time</th><th>Burst Time</th><th>Completion Time</th><th>Turnaround Time</th><th>Waiting Time</th></tr>";

      foreach ($processes as $process) {
        $processName = $process["name"];
        $arrivalTime = $process["arrivalTime"];
        $burstTime = $process["burstTime"];
        $completionTime = isset($completionTimes[$processName]) ? $completionTimes[$processName] : "N/A";
        $turnaroundTime = isset($turnaroundTimes[$processName]) ? $turnaroundTimes[$processName] : "N/A";
        $waitingTime = isset($waitingTimes[$processName]) ? $waitingTimes[$processName] : "N/A";

        echo "<tr><td>$processName</td><td>$arrivalTime</td><td>$burstTime</td><td>$completionTime</td><td>$turnaroundTime</td><td>$waitingTime</td></tr>";
      }

      echo "<tr><th colspan='4'>Average</th><th>" . array_sum($turnaroundTimes) / count($turnaroundTimes) . "</th><th>" . array_sum($waitingTimes) / count($waitingTimes) . "</th></tr>";

      echo "</table>";

      //Ganttcharttable
      echo "<h2>Gantt Chart</h2>";
      echo "<table>";
      echo "<tr><th>Process</th>";

		// process names
		foreach ($ganttChart as $time => $process) {
		echo "<th>$process</th>";
		}
		echo "</tr>";
		echo "<tr><td>Time</td>";
		// time
		foreach ($ganttChart as $time => $process) {
		echo "<td>$time</td>";
		}
		echo "</tr>";


      echo "</table>";
      
      
    }
    ?>
    

  </div>
</body>
</html>
  </div>
</body>
</html>

File: sjf_non_preemptive.py
class SJF:

    def processData(self, no_of_processes):
        process_data = []
        for i in range(no_of_processes):
            temporary = []
            process_id = int(input("Enter Process ID: "))
            arrival_time = int(input(f"Enter Arrival Time for Process {process_id}: "))
            burst_time = int(input(f"Enter Burst Time for Process {process_id}: "))
            temporary.extend([process_id, arrival_time, burst_time, 0])
            process_data.append(temporary)
        self.schedulingProcess(process_data)

    def schedulingProcess(self, process_data):
        start_time = []
        exit_time = []
        s_time = 0
        process_data.sort(key=lambda x: x[1])

        for i in range(len(process_data)):
            ready_queue = []
            temp = []
            normal_queue = []

            for j in range(len(process_data)):
                if (process_data[j][1] <= s_time) and (process_data[j][3] == 0):
                    temp.extend([process_data[j][0], process_data[j][1], process_data[j][2]])
                    ready_queue.append(temp)
                    temp = []
                elif process_data[j][3] == 0:
                    temp.extend([process_data[j][0], process_data[j][1], process_data[j][2]])
                    normal_queue.append(temp)
                    temp = []

            if len(ready_queue) != 0:
                ready_queue.sort(key=lambda x: x[2])
                '''
                Sort the processes according to the Burst Time
                '''
                start_time.append(s_time)
                s_time = s_time + ready_queue[0][2]
                e_time = s_time
                exit_time.append(e_time)
                for k in range(len(process_data)):
                    if process_data[k][0] == ready_queue[0][0]:
                        break
                process_data[k][3] = 1
                process_data[k].append(e_time)

            elif len(ready_queue) == 0:
                if s_time < normal_queue[0][1]:
                    s_time = normal_queue[0][1]
                start_time.append(s_time)
                s_time = s_time + normal_queue[0][2]
                e_time = s_time
                exit_time.append(e_time)
                for k in range(len(process_data)):
                    if process_data[k][0] == normal_queue[0][0]:
                        break
                process_data[k][3] = 1
                process_data[k].append(e_time)

        t_time = self.calculateTurnaroundTime(process_data)  # Call the method after the scheduling loop
        w_time = self.calculateWaitingTime(process_data)
        self.printData(process_data, t_time, w_time)


    def calculateTurnaroundTime(self, process_data):
        total_turnaround_time = 0
        for i in range(len(process_data)):
            turnaround_time = process_data[i][4] - process_data[i][1]
            '''
            turnaround_time = completion_time - arrival_time
            '''
            total_turnaround_time = total_turnaround_time + turnaround_time
            process_data[i].append(turnaround_time)
        average_turnaround_time = total_turnaround_time / len(process_data)
        '''
        average_turnaround_time = total_turnaround_time / no_of_processes
        '''
        return average_turnaround_time


    def calculateWaitingTime(self, process_data):
        total_waiting_time = 0
        for i in range(len(process_data)):
            waiting_time = process_data[i][5] - process_data[i][2]
            '''
            waiting_time = turnaround_time - burst_time
            '''
            total_waiting_time = total_waiting_time + waiting_time
            process_data[i].append(waiting_time)
        average_waiting_time = total_waiting_time / len(process_data)
        '''
        average_waiting_time = total_waiting_time / no_of_processes
        '''
        return average_waiting_time


    def printData(self, process_data, average_turnaround_time, average_waiting_time):
        
        process_data.sort(key=lambda x: x[6])  # Sort based on completion time
        
        print("\nShortest Job First Non Preemptive")

        print("\nGantt Chart")
        ganttWidths = [process_data[i][6] - process_data[i][1] for i in range(len(process_data))]

        print("+", end="")
        for width in ganttWidths:
            print("-" * 4 * width, end="+")
        print()

        print("|", end="")
        for i in range(len(process_data)):
            width = ganttWidths[i]
            process_id = process_data[i][0]
            print(" " * (4 * width - 3), f"P{process_id}", " " * (4 * width - 3), end="|")
        print()

        print("+", end="")
        for width in ganttWidths:
            print("-" * 4 * width, end="+")
        print()

        ganttStartTimes = [0] + [process_data[i][6] for i in range(len(process_data))]
        print(ganttStartTimes[0], end="")
        for i in range(len(process_data)):
            width = ganttWidths[i]
            print(" " * (4 * width), ganttStartTimes[i + 1], end="")
        print()
        prev_exit_time = 0
        for i in range(len(process_data)):
            process_id = process_data[i][0]
            start_time = process_data[i][6]
            end_time = process_data[i][6] + process_data[i][2]

            # Print spaces for idle time (if any) before this process starts
            if start_time > prev_exit_time:
                print(" " * (start_time - prev_exit_time), end="")

            # Print the process ID for its duration
            print("-" * (end_time - start_time), end="")
            prev_exit_time = end_time

        print("\n")
        print("Process_ID  Arrival_Time  Burst_Time      Completed  Completion_Time  Turnaround_Time  Waiting_Time")

        for i in range(len(process_data)):
            for j in range(len(process_data[i])):
                print(process_data[i][j], end=" " * (15 - len(str(process_data[i][j]))))
            print()

        print(f'Average Turnaround Time: {average_turnaround_time}')
        print(f'Average Waiting Time: {average_waiting_time}')



if __name__ == "__main__":
    no_of_processes = int(input("Enter number of processes: "))
    sjf = SJF()
    sjf.processData(no_of_processes)